### PR TITLE
Expose API endpoint for Events KPI

### PR DESCRIPTION
**Pull request by johngian**

### CODE REVIEW

File: remo/events/api/views.py [hunk 1]
```diff
@@ -1,12 +1,24 @@
+import math
+
```
Comment (akatsoulas) [inline remo/events/api/views.py]: Remove this line

File: remo/events/api/views.py [hunk 2]
```diff
@@ -1,12 +1,24 @@
...
+from collections import namedtuple
+from datetime import datetime, timedelta
+
+import django_filters
```
Comment (akatsoulas) [inline remo/events/api/views.py]: move this above rest_framework.views

File: remo/events/api/views.py [hunk 3]
```diff
@@ -18,3 +30,62 @@ def retrieve(self, request, pk):
         serializer = EventDetailedSerializer(event,
                                              context={'request': request})
         return Response(serializer.data)
+
+
+class EventsKPIFilter(django_filters.FilterSet):
```
Comment (akatsoulas) [inline remo/events/api/views.py]: Is there any chance that we will need this for the public serializers?

File: remo/events/tests/test_api.py [hunk 4]
```diff
@@ -67,3 +70,91 @@ def test_get_remo_url(self):
...
+        eq_(response.data['total'], 10)
+
+    @patch('remo.events.api.views.now')
+    def test_quarter(self, mock_now):
+        mock_now.return_value = datetime(2015, 3, 1)
+
+        # Previous quarter
+        start = datetime(2014, 12, 5)
+        end = datetime(2014, 12, 6)
+        EventFactory.create_batch(10, start=start, end=end)
```
Comment (akatsoulas) [inline remo/events/tests/test_api.py]: There's no need to create so many events. You can make your test case with only one event per quarter for example.

File: remo/events/api/views.py [hunk 5]
```diff
@@ -18,3 +30,62 @@ def retrieve(self, request, pk):
...
+        current_q = int(math.ceil(now().month/3.0))
+        q_start = datetime(now().year, current_q, 1)
+        total_q = f.qs.filter(start__gte=q_start).count()
+        total_soq = f.qs.filter(start__lte=q_start).count()
+        try:
+            percentage_quarter = (total-total_soq)/float(total_soq)
+        except ZeroDivisionError:
+            percentage_quarter = 0
+
+            # Week calculations
```
Comment (comzeradd) [inline remo/events/api/views.py]: This comment line seems over-indented

File: remo/events/api/views.py [hunk 6]
```diff
@@ -18,3 +30,62 @@ def retrieve(self, request, pk):
...
+        f = EventsKPIFilter(request.GET, queryset=Event.objects.all())
+        weeks = int(request.QUERY_PARAMS.get('weeks', KPI_WEEKS))
+
+        # Total calculations
+        total = f.qs.count()
+
+        # Quarter calculations
+        current_q = int(math.ceil(now().month/3.0))
+        q_start = datetime(now().year, current_q, 1)
+        total_q = f.qs.filter(start__gte=q_start).count()
```
Comment (akatsoulas) [inline remo/events/api/views.py]: Why don't you just apply a `range` of dates?

File: remo/events/api/views.py (hunk 3 above)
Comment (johngian) [inline remo/events/api/views.py]: This is a subset of the filters required in the public api view. I suggest we implement for the KPIs for now and extend it or revisit it when we tackle the public API filters.

File: remo/events/tests/test_api.py [hunk 7]
```diff
@@ -67,3 +70,91 @@ def test_get_remo_url(self):
...
+
+        # This quarter
+        start = datetime(2015, 1, 5)
+        end = datetime(2015, 1, 6)
+        EventFactory.create_batch(20, start=start, end=end)
```
Comment (akatsoulas) [inline remo/events/tests/test_api.py]: Probably if you add a few events ahead of `this quarter`, they will be included in the response.

File: remo/events/api/views.py (hunk 3 above)
Comment (akatsoulas) [inline remo/events/api/views.py]: :+1:

File: remo/events/api/serializers.py [hunk 8]
```diff
@@ -34,3 +34,13 @@ def get_remo_url(self, obj):
...
+
+
+class EventsKPISerializer(serializers.Serializer):
+    """Serializer for the Events KPI data"""
+    total = serializers.IntegerField()
+    total_quarter = serializers.IntegerField()
+    percentage_quarter = serializers.FloatField()
+    total_week = serializers.IntegerField()
+    percentage_week = serializers.FloatField()
+    total_per_week = serializers.DictField()
```
Comment (comzeradd) [inline remo/events/api/serializers.py]: Can we amend the total_per_week to group results by label. Something like this:

```
"total_per_week": [
        {
            "week": "1",
            "events": 13
        },
        {
            "week": "2",
            "events": 10
        }
]
```

File: remo/events/api/views.py (hunk 6 above)
Comment (johngian) [inline remo/events/api/views.py]: I have applied ``range`` to other references that this comment applied. I also updated the initial queryset in ``f.qs`` filtering only current or past events. We can leave this without any change.

**Comment by johngian** [comment]
@akatsoulas @comzeradd PR updated

**Comment by akatsoulas** [comment]
Since this not going to master and we have the luxury to test it as we proceed with the kpis I would say this is good to go. My only comment is that  I would prefer different naming in some cases eg `f.qs`. FYI I haven't tested this. r+wc from me

**Comment by comzeradd** [comment]
Tested locally along with d3js and it looks good.

r+ :rocket:

**Comment by johngian** [comment]
Thank y'all :heart:

**Pull request merged**